### ISSUE TITLE
`Integer divide by zero` after remote shard removed

### ISSUE THREAD

**Issue by asxdfd** [ClickHouse/ClickHouse#39964]
**Describe what's wrong**

We have a cluster containing 36 shards. After scaling down the cluster, one of the clickhouse servers crashed. We checked the server log and it shows ther server `Received signal Arithmetic exception (8)` and `Integer divide by zero`. Another exception is `DB::Exception: No shard with shard_index=36 (shard36_replica1,shard36_replica2)`. Here the version is 21.3.

**Does it reproduce on recent release?**

Reproduce on version v22.3.6.5.

**How to reproduce**

* ClickHouse server version: v22.3.6.5, build from source
* Cluster settings for reproduce
```
    <clickhouse_remote_servers>
        <all_shard>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>127.0.0.1</host>
                    <port>29910</port>
                    <user>default</user>
                    <password>test</password>
                </replica>
            </shard>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>127.0.0.1</host>
                    <port>29911</port>
                    <user>default</user>
                    <password>test</password>
                </replica>
            </shard>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>127.0.0.1</host>
                    <port>29912</port>
                    <user>default</user>
                    <password>test</password>
                </replica>
            </shard>
        </all_shard>
    </clickhouse_remote_servers>
```
* Besides, I make some modifications to force distributed table write to shard 3.
* `CREATE TABLE` statements for all tables involved
```
CREATE TABLE test1.tt ON CLUSTER all_shard
(
    `a` UInt32,
    `b` UInt32,
    `c` UInt32
)
ENGINE = MergeTree()
ORDER BY a

CREATE TABLE tt_data (a UInt32, b UInt32, c UInt32) ENGINE = MergeTree() ORDER BY a;

CREATE TABLE tt_all ON CLUSTER all_shard AS test1.tt ENGINE = Distributed('all_shard', 'test1', 'tt', rand());
```
* Queries to run that lead to unexpected result
  - insert some data in table `tt_data`
  - `INSERT INTO tt_all SELECT * FROM tt_data;`
  - Stop shard3 before insert completed. That is, file `/path/data/store/uuid/shard3_all_replicas/1.bin` wouldn't be sent to remote shard.
  - Shard1 server shows `DB::NetException: Connection refused (127.0.0.1:29912)`
  - Change the cluster settings for shard1 and shard2, remove shard with port `29912`
  ```
    <clickhouse_remote_servers>
        <all_shard>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>127.0.0.1</host>
                    <port>29910</port>
                    <user>default</user>
                    <password>test</password>
                </replica>
            </shard>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>127.0.0.1</host>
                    <port>29911</port>
                    <user>default</user>
                    <password>test</password>
                </replica>
            </shard>
        </all_shard>
    </clickhouse_remote_servers>
  ```
  - Restart shard1 server, it shows `Received signal Arithmetic exception (8)` and `Integer divide by zero`. Then it is termitated.

**Expected behavior**

Server can run successfully.

**Error message and/or stacktrace**

- For `Connection refused`. This exception do not influence server.
```
2022.08.05 11:14:12.742706 [ 24304 ] {} <Debug> test1.tt_all.DirectoryMonitor: Started processing `/path/workspace/cluster/node1/data/store/7c8/7c8344cd-257e-4ca5-bda3-ee6dc0ca52c3/shard3_all_replicas/1.bin` (1.00 million rows, 11.44 MiB bytes)
2022.08.05 11:14:12.743325 [ 24304 ] {} <Error> test1.tt_all.DirectoryMonitor: Code: 210. DB::NetException: Connection refused (127.0.0.1:29912): While sending /path/workspace/cluster/node1/data/store/7c8/7c8344cd-257e-4ca5-bda3-ee6dc0ca52c3/shard3_all_replicas/1.bin. (NETWORK_ERROR), Stack trace (when copying this message, always include the lines below):

0. ./build/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x17037e6c in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
1. ./build/../src/Common/Exception.cpp:58: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb34297a in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2. ./build/../src/Common/NetException.h:12: DB::Connection::connect(DB::ConnectionTimeouts const&) @ 0x1485848d in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
3. ./build/../src/Client/ConnectionPool.h:92: DB::ConnectionPool::get(DB::ConnectionTimeouts const&, DB::Settings const*, bool) @ 0xb416367 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
4. ./build/../src/Storages/Distributed/DirectoryMonitor.cpp:626: DB::StorageDistributedDirectoryMonitor::processFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x1454c283 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
5. ./build/../contrib/libcxx/include/__tree:182: DB::StorageDistributedDirectoryMonitor::run() @ 0x14548adf in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
6. ./build/../contrib/libcxx/include/functional:0: DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x134722fd in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
7. ./build/../contrib/poco/Foundation/include/Poco/AtomicCounter.h:314: DB::BackgroundSchedulePool::threadFunction() @ 0x13473b87 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
8. ./build/../src/Core/BackgroundSchedulePool.cpp:0: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x134741f8 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
9. ./build/../contrib/libcxx/include/functional:2210: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xb3c583a in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
10. ./build/../contrib/libcxx/include/memory:1655: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0xb3c7bce in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
11. ? @ 0x2b351197f609 in ?
12. __clone @ 0x2b3511ab9133 in ?
 (version 22.3.6.5)
```
- For `Integer divide by zero`. This is a FATAL which will terminate the server.
```
2022.08.05 11:14:31.410516 [ 24544 ] {} <Error> StorageDistributed (tt_all): No shard with shard_index=3 (shard3_all_replicas)
2022.08.05 11:14:31.461124 [ 24637 ] {} <Debug> test1.tt_all.DirectoryMonitor: Started processing `/path/workspace/cluster/node1/data/store/7c8/7c8344cd-257e-4ca5-bda3-ee6dc0ca52c3/shard3_all_replicas/1.bin` (1.00 million rows, 11.44 MiB bytes)
2022.08.05 11:14:31.463991 [ 24689 ] {} <Fatal> BaseDaemon: ########################################
2022.08.05 11:14:31.464086 [ 24689 ] {} <Fatal> BaseDaemon: (version 22.3.6.5, build id: 2DF710A2D8EC385D) (from thread 24637) (no query) Received signal Arithmetic exception (8)
2022.08.05 11:14:31.464113 [ 24689 ] {} <Fatal> BaseDaemon: Integer divide by zero.
2022.08.05 11:14:31.464149 [ 24689 ] {} <Fatal> BaseDaemon: Stack trace: 0x148675bc 0x1454c283 0x14548adf 0x134722fd 0x13473b87 0x134741f8 0xb3c583a 0xb3c7bce 0x2b99b6ad7609 0x2b99b6c11133
2022.08.05 11:14:31.478730 [ 24689 ] {} <Fatal> BaseDaemon: 3. ./build/../src/Client/ConnectionPoolWithFailover.cpp:55: DB::ConnectionPoolWithFailover::get(DB::ConnectionTimeouts const&, DB::Settings const*, bool) @ 0x148675bc in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.516824 [ 24689 ] {} <Fatal> BaseDaemon: 4. ./build/../src/Storages/Distributed/DirectoryMonitor.cpp:626: DB::StorageDistributedDirectoryMonitor::processFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x1454c283 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.553533 [ 24689 ] {} <Fatal> BaseDaemon: 5.1. inlined from ./build/../contrib/libcxx/include/__tree:182: std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>* std::__1::__tree_next_iter<std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>*, std::__1::__tree_node_base<void*>*>(std::__1::__tree_node_base<void*>*)
2022.08.05 11:14:31.553601 [ 24689 ] {} <Fatal> BaseDaemon: 5.2. inlined from ../contrib/libcxx/include/__tree:923: std::__1::__tree_const_iterator<std::__1::__value_type<unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::__tree_node<std::__1::__value_type<unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, void*>*, long>::operator++()
2022.08.05 11:14:31.553634 [ 24689 ] {} <Fatal> BaseDaemon: 5.3. inlined from ../contrib/libcxx/include/map:867: std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::__tree_node<std::__1::__value_type<unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, void*>*, long> >::operator++()
2022.08.05 11:14:31.553661 [ 24689 ] {} <Fatal> BaseDaemon: 5.4. inlined from ../src/Storages/Distributed/DirectoryMonitor.cpp:598: DB::StorageDistributedDirectoryMonitor::processFiles(std::__1::map<unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&)
2022.08.05 11:14:31.553682 [ 24689 ] {} <Fatal> BaseDaemon: 5. ../src/Storages/Distributed/DirectoryMonitor.cpp:437: DB::StorageDistributedDirectoryMonitor::run() @ 0x14548adf in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.560261 [ 24689 ] {} <Fatal> BaseDaemon: 6. ./build/../contrib/libcxx/include/functional:0: DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x134722fd in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.568898 [ 24689 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build/../contrib/poco/Foundation/include/Poco/AtomicCounter.h:314: Poco::AtomicCounter::operator--()
2022.08.05 11:14:31.568967 [ 24689 ] {} <Fatal> BaseDaemon: 7.2. inlined from ../contrib/poco/Foundation/include/Poco/RefCountedObject.h:82: Poco::RefCountedObject::release() const
2022.08.05 11:14:31.568987 [ 24689 ] {} <Fatal> BaseDaemon: 7.3. inlined from ../contrib/poco/Foundation/include/Poco/AutoPtr.h:90: ~AutoPtr
2022.08.05 11:14:31.569004 [ 24689 ] {} <Fatal> BaseDaemon: 7. ../src/Core/BackgroundSchedulePool.cpp:260: DB::BackgroundSchedulePool::threadFunction() @ 0x13473b87 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.577936 [ 24689 ] {} <Fatal> BaseDaemon: 8.1. inlined from ./build/../src/Core/BackgroundSchedulePool.cpp:0: operator()
2022.08.05 11:14:31.578012 [ 24689 ] {} <Fatal> BaseDaemon: 8.2. inlined from ../contrib/libcxx/include/type_traits:3676: decltype(std::__1::forward<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(fp)()) std::__1::__invoke<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'()&>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)
2022.08.05 11:14:31.578040 [ 24689 ] {} <Fatal> BaseDaemon: 8.3. inlined from ../contrib/libcxx/include/__functional_base:348: void std::__1::__invoke_void_return_wrapper<void>::__call<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'()&>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&...)
2022.08.05 11:14:31.578072 [ 24689 ] {} <Fatal> BaseDaemon: 8.4. inlined from ../contrib/libcxx/include/functional:1608: std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'(), void ()>::operator()()
2022.08.05 11:14:31.578092 [ 24689 ] {} <Fatal> BaseDaemon: 8. ../contrib/libcxx/include/functional:2089: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::$_1&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x134741f8 in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.587241 [ 24689 ] {} <Fatal> BaseDaemon: 9.1. inlined from ./build/../contrib/libcxx/include/functional:2210: std::__1::__function::__policy_func<void ()>::operator=(std::nullptr_t)
2022.08.05 11:14:31.587316 [ 24689 ] {} <Fatal> BaseDaemon: 9.2. inlined from ../contrib/libcxx/include/functional:2533: std::__1::function<void ()>::operator=(std::nullptr_t)
2022.08.05 11:14:31.587335 [ 24689 ] {} <Fatal> BaseDaemon: 9. ../src/Common/ThreadPool.cpp:280: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xb3c583a in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.598288 [ 24689 ] {} <Fatal> BaseDaemon: 10.1. inlined from ./build/../contrib/libcxx/include/memory:1655: std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >::reset(std::__1::__thread_struct*)
2022.08.05 11:14:31.598355 [ 24689 ] {} <Fatal> BaseDaemon: 10.2. inlined from ../contrib/libcxx/include/memory:1612: ~unique_ptr
2022.08.05 11:14:31.598374 [ 24689 ] {} <Fatal> BaseDaemon: 10.3. inlined from ../contrib/libcxx/include/tuple:181: ~__tuple_leaf
2022.08.05 11:14:31.598394 [ 24689 ] {} <Fatal> BaseDaemon: 10.4. inlined from ../contrib/libcxx/include/tuple:477: ~tuple
2022.08.05 11:14:31.598419 [ 24689 ] {} <Fatal> BaseDaemon: 10.5. inlined from ../contrib/libcxx/include/memory:1397: std::__1::default_delete<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >::operator()(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>*) const
2022.08.05 11:14:31.598448 [ 24689 ] {} <Fatal> BaseDaemon: 10.6. inlined from ../contrib/libcxx/include/memory:1658: std::__1::unique_ptr<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>, std::__1::default_delete<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> > >::reset(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>*)
2022.08.05 11:14:31.598480 [ 24689 ] {} <Fatal> BaseDaemon: 10.7. inlined from ../contrib/libcxx/include/memory:1612: ~unique_ptr
2022.08.05 11:14:31.598498 [ 24689 ] {} <Fatal> BaseDaemon: 10. ../contrib/libcxx/include/thread:293: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0xb3c7bce in /path/workspace/clickhouse-22.3/clickhouse/build/bin/clickhouse
2022.08.05 11:14:31.598521 [ 24689 ] {} <Fatal> BaseDaemon: 11. ? @ 0x2b99b6ad7609 in ?
2022.08.05 11:14:31.598541 [ 24689 ] {} <Fatal> BaseDaemon: 12. __clone @ 0x2b99b6c11133 in ?
2022.08.05 11:14:31.767877 [ 24689 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 8608DC6E85FAD9DFDA43FC3E504DD410. There is no information about the reference checksum.
```

**Additional context**

Maybe this bug is caused by `ConnectionPoolWithFailover::get()` function
```
if (settings)
    offset = settings->load_balancing_first_offset % nested_pools.size();
```
If `address.shard_index > shards_info.size()` in `StorageDistributedDirectoryMonitor::createPool()`, the `nested_pools.size()` will be 0.

**Comment by tavplubix** [comment]
Fixed in #38029
But seems like it was not backported to 22.3 for some reason

**Comment by tavplubix** [comment]
Hm, no, actually it was: https://github.com/ClickHouse/ClickHouse/pull/35373/commits/420bdfa27510fe18e20f881d74ab66cddc44583d
Upgrade to latest patch-release (currently it's 22.3.**11**.22)
